File: Greedy/MaximumUnits/unit.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    typedef pair<int,int>pip;
    int maximumUnits(vector<vector<int>>& boxTypes, int truckSize) {
        priority_queue<pip>pq;
        int n = boxTypes.size();
        for(int i=0;i<n;i++){
            pip p = {boxTypes[i][1],boxTypes[i][0]};
            pq.push(p);
        }
        int ans = 0;
        while(pq.size() && truckSize>0){
            pip p = pq.top();
            pq.pop();
            int boxes = p.second;
            int units = p.first;
            if(truckSize - boxes < 0){
                boxes = truckSize;
                truckSize = 0;
            }else truckSize -= boxes;
            ans += boxes*units;
        }
        return ans;
    }
};

File: Greedy/MaximumUnits/unit.py
import heapq

class Solution:
    def maximumUnits(self, boxTypes, truckSize):
        pq = []

        n = len(boxTypes)

        for i in range(n):
            units = boxTypes[i][1]
            boxes = boxTypes[i][0]

            heapq.heappush(pq, (-units, boxes))

        ans = 0

        while pq and truckSize > 0:
            units, boxes = heapq.heappop(pq)

            units = -units

            if truckSize - boxes < 0:
                boxes = truckSize
                truckSize = 0
            else:
                truckSize -= boxes

            ans += boxes * units

        return ans